stan/math/rev/core/scoped_chainablestack.hpp

```diff
@@ -0,0 +1,70 @@
+#ifndef STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+#define STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+
+#include <stan/math/rev/core/chainablestack.hpp>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * The AD tape of reverse mode AD is by default stored globally within the
+ * process (or thread). With the ScopedChainableStack class one may execute a
+ * nullary functor with reference to an AD tape which is stored with the
+ * instance of ScopedChainableStack. Example:
+ *
+ * ScopedChainableStack scoped_stack;
+ *
+ * double cgrad = scoped_stack.execute([] {
+ *   var a = 1;
+ *   var b = 4;
+ *   var c = a + b;
+ *   grad();
+ *   return c.adj();
+ * });
+ *
+ * Doing so will not interfere with the process (or thread) AD tape.
+ */
+class ScopedChainableStack {
+  ChainableStack::AutodiffStackStorage local_stack_;
+
+  std::vector<ChainableStack::AutodiffStackStorage*> stack_queue_;
+
+  struct activate_scope {
+    ScopedChainableStack& scoped_stack_;
+
+    explicit activate_scope(ScopedChainableStack& scoped_stack)
+        : scoped_stack_(scoped_stack) {
+      scoped_stack_.stack_queue_.push_back(ChainableStack::instance_);
+      ChainableStack::instance_ = &scoped_stack_.local_stack_;
+    }
+
+    ~activate_scope() {
+      ChainableStack::instance_ = scoped_stack_.stack_queue_.back();
+      scoped_stack_.stack_queue_.pop_back();
+    }
+  };
+
+ public:
+  ScopedChainableStack() = default;
+
+  /**
+   * Execute in the current thread a nullary function and write the AD
+   * tape to local_stack_ of this instance. The function may return
+   * any type.
+   *
+   * @tparam F nullary functor to evaluate
+   * @param f instance of nullary functor
+   * @return Result of evaluated functor
+   */
+  template <typename F>
+  auto execute(F&& f) {
+    activate_scope active_scope(*this);
```

Comment:
To use the `std::vector` of ChainableStacks one scoped stack would need to call execute on itself right? Like:

```
scoped_stack.execute([&] {
  scoped_stack.execute([] {
     ...
   });
 });
```

I think this can just be one ChainableStack pointer.

Reply:
No. Don’t think so. The entire reduce sum stuff is recorded to scoped stacks which themselves are stored as part of a special vari on the main stack.

Reply:
Sure, but I don't get why we need to save `std::vector` s of things in ScopedStack. The `ChainableStack::AutodiffStackStorage` stacks themselves are the vectors of things.

So if we were going to use a new `ChainableStack::AutodiffStackStorage`, we need to:

1. Save the old `ChainableStack::instance_`
2. Allocate a new Autodiffstack
3. Assign it to `ChainableStack::instance_`
4. Do our stuff
5. Replace the old `ChainableStack::instance_`

As long as when we use another stack we take those five steps, then it's fine if other things in step 4 allocate stacks and such -- they're responsible for restoring the `ChainableStack::instance_` on return. I don't think we need `std::vector` s of Stacks in the ScopedChainable thing.

As far as reverse mode, if we want to extend the life of our autodiff stacks to the life of the stack, we need to put them in a vari. In that case I think ScopedChainableStack becomes the vari. Because these varis will contain `std::vector` s, they varis need to be `chainable_alloc` varis.

Reply:
Uh... sorry, I misread. Ignore my last comment. The need for basically a queue of chainablestack pointers is exactly for the case of nesting calls into one another as you exemplify. I think this should be supported by the design. When doing that, you need a queue of what is nested into one another in order to reverse things. Not sure how you would otherwise do it. I don't say that we should do nesting of things, but I also don't see how to avoid it nor do I think we need to rule it out.

Reply:
> you need a queue of what is nested into one another in order to reverse things.

For forward pass, yeah, but all the necessary stuff will just be saved on the CPU stack as it executes.

For reverse pass, you follow the same rules as forward pass in the chain call of the ScopedChainableVari (save the old instance, load up the new one, do stuff, restore the old instance). If during `do stuff` you chain more ScopedChainableVaris, then those are responsible for restoring the state of `ChainableStack::instance_` when they are done.

Reply:
That's right... but the logic I implemented allows to nest execute calls into one anpther *using the same instance*. What you describe is what happens when we use multiple different instances which nest into one another.

Reply:
> calls into one anpther using the same instance

Yeah and so I was trying to figure out where that would happen. This was the example I came up with was something like:

```
ScopedStack stack;
stack.execute([&stack]() {
  stack.execute([] {
    ...
  });
});
```

but that seems weird (you really have to try to reuse the same stack), and you could solve this with multiple stacks doing:
```
ScopedStack stack1;
stack1.execute([]() {
  ScopedStack stack2;
  stack2.execute([]() {
    ...
  });
});
```